# The Land of Poets Challenge


For a digital humanities project you need to display Italian poets by filtering a csv table according to various criteria. This challenge will be only about querying with Polars, which is something you might find convenient to do during exams for quickly understanding datasets content (using Polars will always be optional, you will never be asked to perform complex modifications with it)

You are given a dataset taken from [Wikidata](https://www.wikidata.org/wiki/Wikidata:Main_Page), a project by the Wikimedia foundation which aims to store only machine-readable data, like numbers, strings, and so on interlinked with many references. Each entity in Wikidata has an identifier, for example Dante Alighieri is the entity [Q1067](https://www.wikidata.org/wiki/Q1067) and Florence is [Q2044](https://www.wikidata.org/wiki/Q2044)

Wikidata can be queried using the SPARQL language: the data was obtained with [this query](https://query.wikidata.org/#%23defaultView%3AMap%7B%22hide%22%3A%20%5B%22%3Fcoord%22%5D%7D%0ASELECT%20%3Fsubj%20%3FsubjLabel%20%3Fplace%20%3FplaceLabel%20%3Fcoord%20%3Fbirthyear%0AWHERE%20%7B%0A%20%20%20%3Fsubj%20wdt%3AP106%20wd%3AQ49757%20.%0A%20%20%20%3Fsubj%20wdt%3AP19%20%3Fplace%20.%0A%20%20%20%3Fplace%20wdt%3AP17%20wd%3AQ38%20.%0A%20%20%20%3Fplace%20wdt%3AP625%20%3Fcoord%20.%0A%20%20%20OPTIONAL%20%7B%20%3Fsubj%20wdt%3AP569%20%3Fdob%20%7D%0A%20%20%20BIND%28YEAR%28%3Fdob%29%20as%20%3Fbirthyear%29%0ASERVICE%20wikibase%3Alabel%20%7B%20%20bd%3AserviceParam%20wikibase%3Alanguage%20%22%5BAUTO_LANGUAGE%5D%2Cen%22%20%7D%0A%7D) and downloaded in CSV format (among the many which can be chosen). Even if not necessary for the purposes of the exercise, you are invited to play a bit with the interface, like trying different visualizations (i.e. try select map in the middle-left corner) - or see [other examples](https://www.wikidata.org/wiki/Wikidata:SPARQL_query_service/queries/examples)

In [ ]:
import polars as pl

## Load the dataset

First load the dataset [italian-poets.csv](data/italian-poets.csv) in Polars dataframe `df` 

In [ ]:
df

shape: (3_198, 9)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬──────────┬───────────┐
│ subj      ┆ subjLabel ┆ place     ┆ placeLabe ┆ … ┆ birthyear ┆ latitude  ┆ place_id ┆ subj_id   │
│ ---       ┆ ---       ┆ ---       ┆ l         ┆   ┆ ---       ┆ ---       ┆ ---      ┆ ---       │
│ str       ┆ str       ┆ str       ┆ ---       ┆   ┆ i64       ┆ f64       ┆ str      ┆ str       │
│           ┆           ┆           ┆ str       ┆   ┆           ┆           ┆          ┆           │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪══════════╪═══════════╡
│ http://ww ┆ Aemilius  ┆ http://ww ┆ Verona    ┆ … ┆ null      ┆ 45.438611 ┆ Q2028    ┆ Q8797     │
│ w.wikidat ┆ Macer     ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Gaius     ┆ http://ww ┆ Arezzo    ┆ … ┆ null      ┆ 43.463056 ┆ Q13378   ┆ Q8833     │
│ w.wikidat ┆ Maecenas  ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Michelang ┆ http://ww ┆ Caprese   ┆ … ┆ 1475      ┆ 43.640833 ┆ Q52069   ┆ Q5592     │
│ w.wikidat ┆ elo       ┆ w.wikidat ┆ Michelang ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆ elo       ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Horace    ┆ http://ww ┆ Venosa    ┆ … ┆ -64       ┆ 40.966667 ┆ Q52691   ┆ Q6197     │
│ w.wikidat ┆           ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Sallust   ┆ http://ww ┆ Amiternum ┆ … ┆ -85       ┆ 42.400776 ┆ Q177061  ┆ Q7170     │
│ w.wikidat ┆           ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …        ┆ …         │
│ http://ww ┆ Giovanni  ┆ http://ww ┆ Noto      ┆ … ┆ 1405      ┆ 36.883333 ┆ Q487174  ┆ Q99308713 │
│ w.wikidat ┆ Marrasio  ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Annemarie ┆ http://ww ┆ Bolzano   ┆ … ┆ 1924      ┆ 46.5      ┆ Q6526    ┆ Q10077537 │
│ w.wikidat ┆ Innerebne ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆ 7         │
│ a.org/ent ┆ r         ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Nesti     ┆ http://ww ┆ Florence  ┆ … ┆ 1875      ┆ 43.771389 ┆ Q2044    ┆ Q10077541 │
│ w.wikidat ┆ Lyro      ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆ 0         │
│ a.org/ent ┆ Wollek    ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Lorenza   ┆ http://ww ┆ Bonde

## Tell me more

Show some info about the dataset

In [ ]:
info

shape: (9, 7)
┌────────────┬──────────────┬──────────────┬─────────────┬─────────────┬─────────────┬─────────────┐
│ statistic  ┆ subj         ┆ subjLabel    ┆ place       ┆ placeLabel  ┆ coord       ┆ birthyear   │
│ ---        ┆ ---          ┆ ---          ┆ ---         ┆ ---         ┆ ---         ┆ ---         │
│ str        ┆ str          ┆ str          ┆ str         ┆ str         ┆ str         ┆ f64         │
╞════════════╪══════════════╪══════════════╪═════════════╪═════════════╪═════════════╪═════════════╡
│ count      ┆ 3198         ┆ 3198         ┆ 3198        ┆ 3198        ┆ 3198        ┆ 3070.0      │
│ null_count ┆ 0            ┆ 0            ┆ 0           ┆ 0           ┆ 0           ┆ 128.0       │
│ mean       ┆ null         ┆ null         ┆ null        ┆ null        ┆ null        ┆ 1724.418893 │
│ std        ┆ null         ┆ null         ┆ null        ┆ null        ┆ null        ┆ 373.394174  │
│ min        ┆ http://www.w ┆ Abd          ┆ http://www. ┆ Aci Catena  ┆ Point(10.03 ┆ -629.0      │
│            ┆ ikidata.org/ ┆ ar-Rahman    ┆ wikidata.or ┆             ┆ 3333333 44. ┆             │
│            ┆ entity…      ┆ al-Itrabanis ┆ g/entity…   ┆             ┆ 26666666…   ┆             │
│            ┆              ┆ hi           ┆             ┆             ┆             ┆             │
│ 25%        ┆ null         ┆ null         ┆ null        ┆ null        ┆ null        ┆ 1647.0      │
│ 50%        ┆ null         ┆ null         ┆ null        ┆ null        ┆ null        ┆ 1855.0      │
│ 75%        ┆ null         ┆ null         ┆ null        ┆ null        ┆ null        ┆ 1922.0      │
│ max        ┆ http://www.w ┆ Đuro         ┆ http://www. ┆ ancient     ┆ Point(9.983 ┆ 2000.0      │
│            ┆ ikidata.org/ ┆ Matijašević  ┆ wikidata.or ┆ Syracuse    ┆ 333333 45.3 ┆             │
│            ┆ entity…      ┆              ┆ g/entity…   ┆             ┆ 16666666…   ┆             │
└────────────┴──────────────┴──────────────┴─────────────┴─────────────┴─────────────┴─────────────┘

## Getting in shape

Show the rows and the columns counts:

In [ ]:
print(f"rows: {nr_rows}")
print(f"columns: {nr_cols}")

rows: 3198
columns: 6


## 10 rows 

Display first 10 rows

In [ ]:
first_10

shape: (10, 9)
┌────────────┬────────────┬───────────┬───────────┬───┬───────────┬───────────┬──────────┬─────────┐
│ subj       ┆ subjLabel  ┆ place     ┆ placeLabe ┆ … ┆ birthyear ┆ latitude  ┆ place_id ┆ subj_id │
│ ---        ┆ ---        ┆ ---       ┆ l         ┆   ┆ ---       ┆ ---       ┆ ---      ┆ ---     │
│ str        ┆ str        ┆ str       ┆ ---       ┆   ┆ i64       ┆ f64       ┆ str      ┆ str     │
│            ┆            ┆           ┆ str       ┆   ┆           ┆           ┆          ┆         │
╞════════════╪════════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪══════════╪═════════╡
│ http://www ┆ Aemilius   ┆ http://ww ┆ Verona    ┆ … ┆ null      ┆ 45.438611 ┆ Q2028    ┆ Q8797   │
│ .wikidata. ┆ Macer      ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Gaius      ┆ http://ww ┆ Arezzo    ┆ … ┆ null      ┆ 43.463056 ┆ Q13378   ┆ Q8833   │
│ .wikidata. ┆ Maecenas   ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Michelange ┆ http://ww ┆ Caprese   ┆ … ┆ 1475      ┆ 43.640833 ┆ Q52069   ┆ Q5592   │
│ .wikidata. ┆ lo         ┆ w.wikidat ┆ Michelang ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆ elo       ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Horace     ┆ http://ww ┆ Venosa    ┆ … ┆ -64       ┆ 40.966667 ┆ Q52691   ┆ Q6197   │
│ .wikidata. ┆            ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Sallust    ┆ http://ww ┆ Amiternum ┆ … ┆ -85       ┆ 42.400776 ┆ Q177061  ┆ Q7170   │
│ .wikidata. ┆            ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Ovid       ┆ http://ww ┆ Sulmona   ┆ … ┆ -42       ┆ 42.048025 ┆ Q50157   ┆ Q7198   │
│ .wikidata. ┆            ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Grazia     ┆ http://ww ┆ Nuoro     ┆ … ┆ 1871      ┆ 40.320062 ┆ Q13649   ┆ Q7728   │
│ .wikidata. ┆ Deledda    ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Bronzino   ┆ http://ww ┆ Florence  ┆ … ┆ 1503      ┆ 43.771389 ┆ Q2044    ┆ Q7803   │
│ .wikidata. ┆            ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆         │
│ org/entity ┆            ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆         │
│ …          ┆            ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆         │
│ http://www ┆ Sandra     ┆ http://ww ┆ Rome      ┆ … ┆ 1946      ┆ 41.893056 ┆ Q220     ┆ Q8796   │
│ .wikidata. ┆ Lombardi   ┆ w.wikidat ┆       

## Born in Verona

Display all people born in Verona

In [ ]:
verona_df

shape: (37, 6)
┌──────────────────┬──────────────────┬─────────────────┬────────────┬─────────────────┬───────────┐
│ subj             ┆ subjLabel        ┆ place           ┆ placeLabel ┆ coord           ┆ birthyear │
│ ---              ┆ ---              ┆ ---             ┆ ---        ┆ ---             ┆ ---       │
│ str              ┆ str              ┆ str             ┆ str        ┆ str             ┆ i64       │
╞══════════════════╪══════════════════╪═════════════════╪════════════╪═════════════════╪═══════════╡
│ http://www.wikid ┆ Aemilius Macer   ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ null      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Catullus         ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ -83       │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Girolamo         ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1478      │
│ ata.org/entity…  ┆ Fracastoro       ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Guarino da       ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1374      │
│ ata.org/entity…  ┆ Verona           ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Ippolito         ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1753      │
│ ata.org/entity…  ┆ Pindemonte       ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ …                ┆ …                ┆ …               ┆ …          ┆ …               ┆ …         │
│ http://www.wikid ┆ Ilario Casarotti ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1772      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Girolamo Orti    ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1769      │
│ ata.org/entity…  ┆ Manara           ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Paolo Zazzaroni  ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ null      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Angela Nogarola  ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1380      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Bartolomeo       ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1560      │
│ ata.org/entity…  ┆ Tortoletti       ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
└──────────────────┴──────────────────┴─────────────────┴────────────┴─────────────────┴───────────┘

## How many  people in Verona

Display how many people were born in Verona

In [ ]:
verona_count

37

## Python is everywhere

Show poets born in Catania in the year -500 

* mind the minus
* I swear we did not altered the dataset in any way :-)

In [ ]:
catania_minus_500

shape: (1, 6)
┌────────────────────┬───────────┬────────────────────┬────────────┬───────────────────┬───────────┐
│ subj               ┆ subjLabel ┆ place              ┆ placeLabel ┆ coord             ┆ birthyear │
│ ---                ┆ ---       ┆ ---                ┆ ---        ┆ ---               ┆ ---       │
│ str                ┆ str       ┆ str                ┆ str        ┆ str               ┆ i64       │
╞════════════════════╪═══════════╪════════════════════╪════════════╪═══════════════════╪═══════════╡
│ http://www.wikidat ┆ Python of ┆ http://www.wikidat ┆ Catania    ┆ Point(15.08726944 ┆ -500      │
│ a.org/entity…      ┆ Catana    ┆ a.org/entity…      ┆            ┆ 4 37.50266944…    ┆           │
└────────────────────┴───────────┴────────────────────┴────────────┴───────────────────┴───────────┘

## Verona after 1500

Display all people born in Verona after the year 1500

In [ ]:
verona_after_1500

shape: (29, 6)
┌──────────────────┬──────────────────┬─────────────────┬────────────┬─────────────────┬───────────┐
│ subj             ┆ subjLabel        ┆ place           ┆ placeLabel ┆ coord           ┆ birthyear │
│ ---              ┆ ---              ┆ ---             ┆ ---        ┆ ---             ┆ ---       │
│ str              ┆ str              ┆ str             ┆ str        ┆ str             ┆ i64       │
╞══════════════════╪══════════════════╪═════════════════╪════════════╪═════════════════╪═══════════╡
│ http://www.wikid ┆ Ippolito         ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1753      │
│ ata.org/entity…  ┆ Pindemonte       ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Aleardo Aleardi  ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1812      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Cristina Ali     ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1973      │
│ ata.org/entity…  ┆ Farah            ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Francesco        ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1675      │
│ ata.org/entity…  ┆ Scipione,        ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆ marchese d…      ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Marco Antonio    ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1750      │
│ ata.org/entity…  ┆ Zucchi           ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ …                ┆ …                ┆ …               ┆ …          ┆ …               ┆ …         │
│ http://www.wikid ┆ Giambattista     ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1747      │
│ ata.org/entity…  ┆ Mutinelli        ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Pietro Caliari   ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1841      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Ilario Casarotti ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1772      │
│ ata.org/entity…  ┆                  ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Girolamo Orti    ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1769      │
│ ata.org/entity…  ┆ Manara           ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
│ http://www.wikid ┆ Bartolomeo       ┆ http://www.wiki ┆ Verona     ┆ Point(10.992777 ┆ 1560      │
│ ata.org/entity…  ┆ Tortoletti       ┆ data.org/entity ┆            ┆ 777             ┆           │
│                  ┆                  ┆ …               ┆            ┆ 45.43861111…    ┆           │
└──────────────────┴──────────────────┴─────────────────┴────────────┴─────────────────┴───────────┘

## First Antonio

Display all people with Antonio as first name

In [ ]:
first_antonios

shape: (85, 6)
┌─────────────────┬─────────────────┬────────────────┬────────────────┬────────────────┬───────────┐
│ subj            ┆ subjLabel       ┆ place          ┆ placeLabel     ┆ coord          ┆ birthyear │
│ ---             ┆ ---             ┆ ---            ┆ ---            ┆ ---            ┆ ---       │
│ str             ┆ str             ┆ str            ┆ str            ┆ str            ┆ i64       │
╞═════════════════╪═════════════════╪════════════════╪════════════════╪════════════════╪═══════════╡
│ http://www.wiki ┆ Antonio Bonfini ┆ http://www.wik ┆ Ancona         ┆ Point(13.51666 ┆ 1427      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 6666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 43.61666666…   ┆           │
│ http://www.wiki ┆ Antonio Bonfini ┆ http://www.wik ┆ Patrignone     ┆ Point(13.60926 ┆ 1427      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 42.98027)      ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Ferrara        ┆ Point(11.61986 ┆ 1463      │
│ data.org/entity ┆ Tebaldeo        ┆ idata.org/enti ┆                ┆ 5 44.835297)   ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Vicenza        ┆ Point(11.55    ┆ 1842      │
│ data.org/entity ┆ Fogazzaro       ┆ idata.org/enti ┆                ┆ 45.55)         ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Lecco          ┆ Point(9.4      ┆ 1824      │
│ data.org/entity ┆ Ghislanzoni     ┆ idata.org/enti ┆                ┆ 45.85)         ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ …               ┆ …               ┆ …              ┆ …              ┆ …              ┆ …         │
│ http://www.wiki ┆ Antonio Bruni   ┆ http://www.wik ┆ Manduria       ┆ Point(17.63416 ┆ 1593      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 6666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 40.40277777…   ┆           │
│ http://www.wiki ┆ Antonio Decio   ┆ http://www.wik ┆ Orte           ┆ Point(12.38611 ┆ 1560      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 1111           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.46027777…   ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Vasto          ┆ Point(14.70821 ┆ 1770      │
│ data.org/entity ┆ Rossetti        ┆ idata.org/enti ┆                ┆ 9444           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.11158888…   ┆           │
│ http://www.wiki ┆ Antonio Ricci   ┆ http://www.wik ┆ Guardiagrele   ┆ Point(14.22159 ┆ 1952      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 1666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.18922222…   ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Ponte di Piave ┆ Point(12.46666 ┆ 1777      │
│ data.org/entity ┆ Gasparinetti    ┆ idata.org/enti ┆                ┆ 6666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 45.71666666…   ┆           │
└─────────────────┴─────────────────┴────────────────┴────────────────┴────────────────┴───────────┘

## Some Antonio

Display all people with Antonio as one of the names (so also include `'Paolo Antonio Rolli'`)

In [ ]:
some_antonio

shape: (110, 6)
┌─────────────────┬─────────────────┬────────────────┬────────────────┬────────────────┬───────────┐
│ subj            ┆ subjLabel       ┆ place          ┆ placeLabel     ┆ coord          ┆ birthyear │
│ ---             ┆ ---             ┆ ---            ┆ ---            ┆ ---            ┆ ---       │
│ str             ┆ str             ┆ str            ┆ str            ┆ str            ┆ i64       │
╞═════════════════╪═════════════════╪════════════════╪════════════════╪════════════════╪═══════════╡
│ http://www.wiki ┆ Antonio Bonfini ┆ http://www.wik ┆ Ancona         ┆ Point(13.51666 ┆ 1427      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 6666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 43.61666666…   ┆           │
│ http://www.wiki ┆ Antonio Bonfini ┆ http://www.wik ┆ Patrignone     ┆ Point(13.60926 ┆ 1427      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 42.98027)      ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Michelangelo    ┆ http://www.wik ┆ Ferrara        ┆ Point(11.61986 ┆ 1912      │
│ data.org/entity ┆ Antonioni       ┆ idata.org/enti ┆                ┆ 5 44.835297)   ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Ferrara        ┆ Point(11.61986 ┆ 1463      │
│ data.org/entity ┆ Tebaldeo        ┆ idata.org/enti ┆                ┆ 5 44.835297)   ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Vicenza        ┆ Point(11.55    ┆ 1842      │
│ data.org/entity ┆ Fogazzaro       ┆ idata.org/enti ┆                ┆ 45.55)         ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ …               ┆ …               ┆ …              ┆ …              ┆ …              ┆ …         │
│ http://www.wiki ┆ Carlo Antonio   ┆ http://www.wik ┆ Salò           ┆ Point(10.53333 ┆ 1637      │
│ data.org/entity ┆ Bertelli        ┆ idata.org/enti ┆                ┆ 3333 45.6)     ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Antonio Decio   ┆ http://www.wik ┆ Orte           ┆ Point(12.38611 ┆ 1560      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 1111           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.46027777…   ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Vasto          ┆ Point(14.70821 ┆ 1770      │
│ data.org/entity ┆ Rossetti        ┆ idata.org/enti ┆                ┆ 9444           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.11158888…   ┆           │
│ http://www.wiki ┆ Antonio Ricci   ┆ http://www.wik ┆ Guardiagrele   ┆ Point(14.22159 ┆ 1952      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 1666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.18922222…   ┆           │
│ http://www.wiki ┆ Antonio         ┆ http://www.wik ┆ Ponte di Piave ┆ Point(12.46666 ┆ 1777      │
│ data.org/entity ┆ Gasparinetti    ┆ idata.org/enti ┆                ┆ 6666           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 45.71666666…   ┆           │
└─────────────────┴─────────────────┴────────────────┴────────────────┴────────────────┴───────────┘

## Cesares during 1800

Display all people named Cesare who were born in the 1800s

In [ ]:
cesares_1800s

shape: (4, 6)
┌─────────────────┬─────────────────┬────────────────┬────────────────┬────────────────┬───────────┐
│ subj            ┆ subjLabel       ┆ place          ┆ placeLabel     ┆ coord          ┆ birthyear │
│ ---             ┆ ---             ┆ ---            ┆ ---            ┆ ---            ┆ ---       │
│ str             ┆ str             ┆ str            ┆ str            ┆ str            ┆ i64       │
╞═════════════════╪═════════════════╪════════════════╪════════════════╪════════════════╪═══════════╡
│ http://www.wiki ┆ Cesare Meano    ┆ http://www.wik ┆ Turin          ┆ Point(7.7      ┆ 1899      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆                ┆ 45.066666666)  ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆                ┆           │
│ http://www.wiki ┆ Cesare          ┆ http://www.wik ┆ Verona         ┆ Point(10.99277 ┆ 1808      │
│ data.org/entity ┆ Betteloni       ┆ idata.org/enti ┆                ┆ 7777           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 45.43861111…   ┆           │
│ http://www.wiki ┆ Cesare De Titta ┆ http://www.wik ┆ Sant'Eusanio   ┆ Point(14.33333 ┆ 1862      │
│ data.org/entity ┆                 ┆ idata.org/enti ┆ del Sangro     ┆ 3333           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 42.16666666…   ┆           │
│ http://www.wiki ┆ Cesare          ┆ http://www.wik ┆ Rome           ┆ Point(12.48277 ┆ 1858      │
│ data.org/entity ┆ Pascarella      ┆ idata.org/enti ┆                ┆ 7777           ┆           │
│ …               ┆                 ┆ ty…            ┆                ┆ 41.89305555…   ┆           │
└─────────────────┴─────────────────┴────────────────┴────────────────┴────────────────┴───────────┘

## The old ones

Show poets in year of birth order, excluding null values from the result.

**HINT**: see [is_not_null()](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.is_not_null.html)

In [ ]:
oldest_to_youngest

shape: (3_070, 6)
┌──────────────────┬─────────────┬─────────────────┬─────────────────┬─────────────────┬───────────┐
│ subj             ┆ subjLabel   ┆ place           ┆ placeLabel      ┆ coord           ┆ birthyear │
│ ---              ┆ ---         ┆ ---             ┆ ---             ┆ ---             ┆ ---       │
│ str              ┆ str         ┆ str             ┆ str             ┆ str             ┆ i64       │
╞══════════════════╪═════════════╪═════════════════╪═════════════════╪═════════════════╪═══════════╡
│ http://www.wikid ┆ Stesichorus ┆ http://www.wiki ┆ Gioia Tauro     ┆ Point(15.9      ┆ -629      │
│ ata.org/entity…  ┆             ┆ data.org/entity ┆                 ┆ 38.433333333)   ┆           │
│                  ┆             ┆ …               ┆                 ┆                 ┆           │
│ http://www.wikid ┆ Ibycus      ┆ http://www.wiki ┆ Reggio Calabria ┆ Point(15.65     ┆ -600      │
│ ata.org/entity…  ┆             ┆ data.org/entity ┆                 ┆ 38.114438888)   ┆           │
│                  ┆             ┆ …               ┆                 ┆                 ┆           │
│ http://www.wikid ┆ Theognis of ┆ http://www.wiki ┆ Megara Hyblaea  ┆ Point(15.181944 ┆ -569      │
│ ata.org/entity…  ┆ Megara      ┆ data.org/entity ┆                 ┆ 44 37.20388889) ┆           │
│                  ┆             ┆ …               ┆                 ┆                 ┆           │
│ http://www.wikid ┆ Parmenides  ┆ http://www.wiki ┆ Velia           ┆ Point(15.154444 ┆ -514      │
│ ata.org/entity…  ┆             ┆ data.org/entity ┆                 ┆ 444             ┆           │
│                  ┆             ┆ …               ┆                 ┆ 40.15944444…    ┆           │
│ http://www.wikid ┆ Python of   ┆ http://www.wiki ┆ Catania         ┆ Point(15.087269 ┆ -500      │
│ ata.org/entity…  ┆ Catana      ┆ data.org/entity ┆                 ┆ 444             ┆           │
│                  ┆             ┆ …               ┆                 ┆ 37.50266944…    ┆           │
│ …                ┆ …           ┆ …               ┆ …               ┆ …               ┆ …         │
│ http://www.wikid ┆ Giacomo     ┆ http://www.wiki ┆ Piombino        ┆ Point(10.52212  ┆ 1985      │
│ ata.org/entity…  ┆ Panicucci   ┆ data.org/entity ┆                 ┆ 42.934824)      ┆           │
│                  ┆             ┆ …               ┆                 ┆                 ┆           │
│ http://www.wikid ┆ Giovanni    ┆ http://www.wiki ┆ Turin           ┆ Point(7.7       ┆ 1986      │
│ ata.org/entity…  ┆ Bertoglio   ┆ data.org/entity ┆                 ┆ 45.066666666)   ┆           │
│                  ┆             ┆ …               ┆                 ┆                 ┆           │
│ http://www.wikid ┆ Gio Evan    ┆ http://www.wiki ┆ Molfetta        ┆ Point(16.6      ┆ 1988      │
│ ata.org/entity…  ┆             ┆ data.org/entity ┆                 ┆ 41.2)           ┆           │
│                  ┆             ┆ …               ┆                 ┆                 ┆           │
│ http://www.wikid ┆ Q14922292   ┆ http://www.wiki ┆ Battipaglia     ┆ Point(14.983333 ┆ 1991      │
│ ata.org/entity…  ┆             ┆ data.org/entity ┆                 ┆ 333             ┆           │
│                  ┆             ┆ …               ┆                 ┆ 40.61666666…    ┆           │
│ http://www.wikid ┆ D.B.P.I.T.  ┆ http://www.wiki ┆ Rome            ┆ Point(12.482777 ┆ 2000      │
│ ata.org/entity…  ┆             ┆ data.org/entity ┆                 ┆ 777             ┆           │
│                  ┆             ┆ …               ┆                 ┆ 41.89305555…    ┆           │
└──────────────────┴─────────────┴─────────────────┴─────────────────┴─────────────────┴───────────┘

## Cities of poets

Find the 5 cities with most poets, sorted from most to least.

* use `group_by` and `sort` methods

In [ ]:
poets_cities.head()

shape: (5, 2)
┌────────────┬─────┐
│ placeLabel ┆ len │
│ ---        ┆ --- │
│ str        ┆ u32 │
╞════════════╪═════╡
│ Rome       ┆ 198 │
│ Florence   ┆ 165 │
│ Milan      ┆ 121 │
│ Naples     ┆ 113 │
│ Venice     ┆ 94  │
└────────────┴─────┘

## Most duplicated poets

Find first 8 duplicated poets

In [ ]:
most_duped

shape: (8, 2)
┌───────────────────────┬─────┐
│ subjLabel             ┆ len │
│ ---                   ┆ --- │
│ str                   ┆ u32 │
╞═══════════════════════╪═════╡
│ Giambattista Andreini ┆ 4   │
│ Sosiphanes            ┆ 4   │
│ Eliseo Calenzio       ┆ 4   │
│ Alojz Rebula          ┆ 4   │
│ Giovanni della Casa   ┆ 3   │
│ Aulus Gellius         ┆ 3   │
│ Tommaso Grossi        ┆ 3   │
│ Giuseppe Carpani      ┆ 3   │
└───────────────────────┴─────┘

## All duplicated poets

Print the number of all duplicated poets

In [ ]:
print(f"There are {dups} duplicated poets")

There are 118 duplicated poets


## Northern poets

Find all the poets born north of a given `town`. In other words, look for town latitude (the second coordinate in `coords`), print it,  and then filter the table.

- **DO NOT** put constants like `46.5` in your code!
- Start by extracting the latitudes as a new column.
- **NOTE**: `coord` column holds just simple strings!

In [ ]:
northern_poets

shape: (14, 7)
┌──────────────┬──────────────┬──────────────┬──────────────┬──────────────┬───────────┬───────────┐
│ subj         ┆ subjLabel    ┆ place        ┆ placeLabel   ┆ coord        ┆ birthyear ┆ latitude  │
│ ---          ┆ ---          ┆ ---          ┆ ---          ┆ ---          ┆ ---       ┆ ---       │
│ str          ┆ str          ┆ str          ┆ str          ┆ str          ┆ i64       ┆ f64       │
╞══════════════╪══════════════╪══════════════╪══════════════╪══════════════╪═══════════╪═══════════╡
│ http://www.w ┆ Oswald von   ┆ http://www.w ┆ Schöneck     ┆ Point(11.847 ┆ 1377      ┆ 46.820264 │
│ ikidata.org/ ┆ Wolkenstein  ┆ ikidata.org/ ┆ Castle       ┆ 977          ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆ 46.820264)   ┆           ┆           │
│ http://www.w ┆ Simon        ┆ http://www.w ┆ Val Müstair  ┆ Point(10.390 ┆ 1511      ┆ 46.60566  │
│ ikidata.org/ ┆ Lemnius      ┆ ikidata.org/ ┆              ┆ 09 46.60566) ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆              ┆           ┆           │
│ http://www.w ┆ Simon        ┆ http://www.w ┆ Val Müstair  ┆ Point(10.42  ┆ 1511      ┆ 46.6      │
│ ikidata.org/ ┆ Lemnius      ┆ ikidata.org/ ┆              ┆ 46.6)        ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆              ┆           ┆           │
│ http://www.w ┆ Mary de      ┆ http://www.w ┆ Brixen       ┆ Point(11.65  ┆ 1925      ┆ 46.716667 │
│ ikidata.org/ ┆ Rachewiltz   ┆ ikidata.org/ ┆              ┆ 46.716666666 ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆ )            ┆           ┆           │
│ http://www.w ┆ Ignaz        ┆ http://www.w ┆ Merano       ┆ Point(11.163 ┆ 1825      ┆ 46.668889 │
│ ikidata.org/ ┆ Vincenz      ┆ ikidata.org/ ┆              ┆ 888888       ┆           ┆           │
│ entity…      ┆ Zingerle     ┆ entity…      ┆              ┆ 46.66888888… ┆           ┆           │
│ …            ┆ …            ┆ …            ┆ …            ┆ …            ┆ …         ┆ …         │
│ http://www.w ┆ Karl Ziegler ┆ http://www.w ┆ St. Martin   ┆ Point(11.227 ┆ 1812      ┆ 46.78392  │
│ ikidata.org/ ┆              ┆ ikidata.org/ ┆ in Passeier  ┆ 27 46.78392) ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆              ┆           ┆           │
│ http://www.w ┆ Fedele       ┆ http://www.w ┆ Sëlva        ┆ Point(11.760 ┆ 1850      ┆ 46.55472  │
│ ikidata.org/ ┆ Demetz       ┆ ikidata.org/ ┆              ┆ 38 46.55472) ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆              ┆           ┆           │
│ http://www.w ┆ Roberta      ┆ http://www.w ┆ Val Badia    ┆ Point(11.899 ┆ 1970      ┆ 46.68361  │
│ ikidata.org/ ┆ Dapunt       ┆ ikidata.org/ ┆              ┆ 17 46.68361) ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆              ┆           ┆           │
│ http://www.w ┆ Roberta      ┆ http://www.w ┆ Val Badia    ┆ Point(11.9   ┆ 1970      ┆ 46.683333 │
│ ikidata.org/ ┆ Dapunt       ┆ ikidata.org/ ┆              ┆ 46.683333333 ┆           ┆           │
│ entity…      ┆              ┆ entity…      ┆              ┆ )            ┆           ┆           │
│ http://www.w ┆ Anna         ┆ http://www.w ┆ Brixen       ┆ Point(11.65  ┆ 1967      ┆ 46.716667 │
│ ikidata.org/ ┆ Katharina    ┆ ikidata.org/ ┆              ┆ 46.716666666 ┆           ┆           │
│ entity…      ┆ Mair         ┆ entity…      ┆              ┆ )            ┆           ┆           │
└──────────────┴──────────────┴──────────────┴──────────────┴──────────────┴───────────┴───────────┘

## Papers please

Extract subject id (i.e. `Q8797`) and place id (i.e. `Q2028`) and MODIFY `df` by putting them into two new columns `subj_id` and `place_id`

In [ ]:
df

shape: (3_198, 9)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬──────────┬───────────┐
│ subj      ┆ subjLabel ┆ place     ┆ placeLabe ┆ … ┆ birthyear ┆ latitude  ┆ place_id ┆ subj_id   │
│ ---       ┆ ---       ┆ ---       ┆ l         ┆   ┆ ---       ┆ ---       ┆ ---      ┆ ---       │
│ str       ┆ str       ┆ str       ┆ ---       ┆   ┆ i64       ┆ f64       ┆ str      ┆ str       │
│           ┆           ┆           ┆ str       ┆   ┆           ┆           ┆          ┆           │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪══════════╪═══════════╡
│ http://ww ┆ Aemilius  ┆ http://ww ┆ Verona    ┆ … ┆ null      ┆ 45.438611 ┆ Q2028    ┆ Q8797     │
│ w.wikidat ┆ Macer     ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Gaius     ┆ http://ww ┆ Arezzo    ┆ … ┆ null      ┆ 43.463056 ┆ Q13378   ┆ Q8833     │
│ w.wikidat ┆ Maecenas  ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Michelang ┆ http://ww ┆ Caprese   ┆ … ┆ 1475      ┆ 43.640833 ┆ Q52069   ┆ Q5592     │
│ w.wikidat ┆ elo       ┆ w.wikidat ┆ Michelang ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆ elo       ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Horace    ┆ http://ww ┆ Venosa    ┆ … ┆ -64       ┆ 40.966667 ┆ Q52691   ┆ Q6197     │
│ w.wikidat ┆           ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Sallust   ┆ http://ww ┆ Amiternum ┆ … ┆ -85       ┆ 42.400776 ┆ Q177061  ┆ Q7170     │
│ w.wikidat ┆           ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …        ┆ …         │
│ http://ww ┆ Giovanni  ┆ http://ww ┆ Noto      ┆ … ┆ 1405      ┆ 36.883333 ┆ Q487174  ┆ Q99308713 │
│ w.wikidat ┆ Marrasio  ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆           │
│ a.org/ent ┆           ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Annemarie ┆ http://ww ┆ Bolzano   ┆ … ┆ 1924      ┆ 46.5      ┆ Q6526    ┆ Q10077537 │
│ w.wikidat ┆ Innerebne ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆ 7         │
│ a.org/ent ┆ r         ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Nesti     ┆ http://ww ┆ Florence  ┆ … ┆ 1875      ┆ 43.771389 ┆ Q2044    ┆ Q10077541 │
│ w.wikidat ┆ Lyro      ┆ w.wikidat ┆           ┆   ┆           ┆           ┆          ┆ 0         │
│ a.org/ent ┆ Wollek    ┆ a.org/ent ┆           ┆   ┆           ┆           ┆          ┆           │
│ ity…      ┆           ┆ ity…      ┆           ┆   ┆           ┆           ┆          ┆           │
│ http://ww ┆ Lorenza   ┆ http://ww ┆ Bonde

## Unknown poets 

Find all the ids of nameless poets and put them in **a python list**.

* **DO NOT** use loops
* **NOTE** a Series object  from the point of view of Python is just a sequence

In [ ]:
unknown_poets

['Q4360247',
 'Q14922292',
 'Q19130448',
 'Q21207901',
 'Q19984452',
 'Q21209119',
 'Q21282215',
 'Q23673492',
 'Q29049430',
 'Q29052339',
 'Q31763467',
 'Q28465822',
 'Q48809843',
 'Q27553577',
 'Q48811051',
 'Q48861610',
 'Q55441810',
 'Q47468550',
 'Q50327630',
 'Q50330028',
 'Q55897192',
 'Q65019765',
 'Q51845316',
 'Q60838260',
 'Q64433131',
 'Q71684946',
 'Q93338246',
 'Q59187521',
 'Q61136330',
 'Q61450547',
 'Q52107491',
 'Q61790603',
 'Q61791394',
 'Q61827513',
 'Q61895377',
 'Q59851133',
 'Q59851150',
 'Q62066746',
 'Q66736238',
 'Q66921487',
 'Q85421610',
 'Q61080035',
 'Q87068357',
 'Q64031897',
 'Q64167386',
 'Q64364409',
 'Q69818426',
 'Q64512266',
 'Q78162153',
 'Q78499894',
 'Q88264630',
 'Q89674973',
 'Q94998318',
 'Q94325725',
 'Q84138681',
 'Q98102965',
 'Q80705985',
 'Q81100287',
 'Q81738068',
 'Q83643534',
 'Q83808244',
 'Q96097742',
 'Q96245247',
 'Q96248786',
 'Q95485499',
 'Q99196947']

## Better unknown poets

Find all the ids, the birthplace and birthdate of nameless poets born after year 0, and put them in **a python list of dictionaries**.

* **DO NOT** use loops nor list comprehensions

In [ ]:
better_unknown_poets

[{'subj_id': 'Q4360247', 'birthyear': 1907, 'placeLabel': 'Rome'},
 {'subj_id': 'Q14922292', 'birthyear': 1991, 'placeLabel': 'Battipaglia'},
 {'subj_id': 'Q19130448', 'birthyear': 1492, 'placeLabel': 'Vicenza'},
 {'subj_id': 'Q21209119', 'birthyear': 1711, 'placeLabel': 'Giuliano Teatino'},
 {'subj_id': 'Q21282215', 'birthyear': 1590, 'placeLabel': 'Palermo'},
 {'subj_id': 'Q29052339', 'birthyear': 1733, 'placeLabel': 'Taranto'},
 {'subj_id': 'Q31763467', 'birthyear': 1755, 'placeLabel': 'Caltanissetta'},
 {'subj_id': 'Q28465822', 'birthyear': 1700, 'placeLabel': 'Orvieto'},
 {'subj_id': 'Q27553577', 'birthyear': 1250, 'placeLabel': 'Cavriana'},
 {'subj_id': 'Q48811051', 'birthyear': 1550, 'placeLabel': 'Roccabernarda'},
 {'subj_id': 'Q48861610', 'birthyear': 1452, 'placeLabel': 'Vittorio Veneto'},
 {'subj_id': 'Q55441810', 'birthyear': 1844, 'placeLabel': 'Vernio'},
 {'subj_id': 'Q47468550', 'birthyear': 1607, 'placeLabel': 'Florence'},
 {'subj_id': 'Q50327630', 'birthyear': 1700, 'p